### PR TITLE
Add parser option enable_options_value_normalization

### LINKED ISSUE #10853: Stop changing the case for COPY TO option values
### Describe the bug

Currently, we're changing the case of every option value passed with the COPY statement:

https://github.com/apache/datafusion/blob/3773fb7fb54419f889e7d18b73e9eb48069eb08e/datafusion/sql/src/statement.rs#L891-L900

This causes an issue in cases where the option values for some external integrations might be case sensitive (eg. like access keys).

@xinlifoobar found this while working on adding an integration to huggingface in https://github.com/apache/datafusion/pull/10792.

See previous discussion here: https://github.com/apache/datafusion/pull/9723#discussion_r1632879391

### To Reproduce

_No response_

### Expected behavior

There should be some way to pass option values as is for certain integrations.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xinlifoobar**
## Which issue does this PR close?

Closes #10853

### CODE REVIEW

**Comment by berkaysynnada** [comment]
Thanks @xinlifoobar working on this issue. 

What I suggest is that we can let people to define their own normalizations. Rather than directing all values to lowercase, people can declare their own normalization methods for specific option keys. If we choose to do that in this PR, I believe the normalization process should be done during the assignment of table options.

**Comment by xinlifoobar** [comment]
> Thanks @xinlifoobar working on this issue.
> 
> What I suggest is that we can let people to define their own normalizations. Rather than directing all values to lowercase, people can declare their own normalization methods for specific option keys. If we choose to do that in this PR, I believe the normalization process should be done during the assignment of table options.

The APIs that could be used in the `datafusion-sql` crate are limited. Do you think it is a good idea to add a method, e.g., `get_options_normalized_function` a good idea?

https://github.com/apache/datafusion/blob/2f02c43b6283472c7c620c68819050128a7e8544/datafusion/expr/src/planner.rs#L33-L83

**Comment by berkaysynnada** [comment]
> > Thanks @xinlifoobar working on this issue.
> > What I suggest is that we can let people to define their own normalizations. Rather than directing all values to lowercase, people can declare their own normalization methods for specific option keys. If we choose to do that in this PR, I believe the normalization process should be done during the assignment of table options.
> 
> The APIs that could be used in the `datafusion-sql` crate are limited. Do you think it is a good idea to add a method, e.g., `get_options_normalized_function` a good idea?

`SqlParserOptions` is accessible, what do you think about adding it there? But I'm not sure if we can support dynamic method registration. My idea is to provide hard-coded functions like the one below, but I'm uncertain where to bind it:

```
lazy_static! {
    static ref NORMALIZATION_MAP: Mutex<HashMap<String, fn(String) -> String>> = {
        let mut m = HashMap::new();
        m.insert("hf.user_access_token".to_string(), lowercase as fn(String) -> String);
        // Other normalizations
        Mutex::new(m)
    };
}

fn lowercase(s: String) -> String {
    s.to_lowercase()
}
// Other normalizations

```
If the key is not found, use the default normalization or do nothing (according to your `enable_options_value_normalization`).

**Comment by xinlifoobar** [comment]
> But

IMO it won't be a good idea... I thought the *configs are static values. What should you do if you would keep a config file for function values?

**Comment by alamb** [comment]
> What I suggest is that we can let people to define their own normalizations. Rather than directing all values to lowercase, people can declare their own normalization methods for specific option keys. If we choose to do that in this PR, I believe the normalization process should be done during the assignment of table options.

I wonder what is the usecase for a user defined normalization? It may be more complicated to code

I like the approach proposed in this PR. As I understand it if follows what we do for SQL:  by default apply some hard coded normalization (e.g. to lower_case) and let users opt out of that normalization either by:
1. Setting a config value like `datafusion.sql_parser.enable_ident_normalization` 
2. Enclose the value in `"`

File: datafusion/sql/tests/sql_integration.rs
Comment (xinlifoobar) [inline datafusion/sql/tests/sql_integration.rs:167]: The display trait is not implemented for CreateExternalTable, maybe there could be a follow-up task.

File: datafusion/sql/src/statement.rs
Comment (xinlifoobar) [inline datafusion/sql/src/statement.rs:1011]: This check is only enabled for `CreateExternalTable` but not `CopyTo`.

File: datafusion/sql/src/statement.rs
Comment (xinlifoobar) [inline datafusion/sql/src/statement.rs:1154]: Here the value is not normalized at all.

**Comment by berkaysynnada** [comment]
Hi again @xinlifoobar. The idea in my mind was something like this: https://github.com/synnada-ai/datafusion-upstream/commit/341b48458c09f25d2a5873fcee2fce18eb0e1097.

I haven't been able to work on it much, and it can't be compiled right now, but you can understand what I mean as an idea. By removing the lowercase transformations we made while parsing the statement, just like in this PR, it seems that we can easily normalize the fields we want in the desired way.

cc @alamb

**Comment by xinlifoobar** [comment]
> > Hi again @xinlifoobar. The idea in my mind was something like this: [synnada-ai@341b484](https://github.com/synnada-ai/datafusion-upstream/commit/341b48458c09f25d2a5873fcee2fce18eb0e1097).
> > I haven't been able to work on it much, and it can't be compiled right now, but you can understand what I mean as an idea. By removing the lowercase transformations we made while parsing the statement, just like in this PR, it seems that we can easily normalize the fields we want in the desired way.
> > cc @alamb
> 
> Hi, I got the idea and am okay to make the change if we agree. The question is still the normalize go beyond the config itself. I am still unsure about this design because the normalize function is just for a small set of configs. in the future, if we have multiple functions for different purposes, the macro might become huge and difficult to read.

Hey @berkaysynnada @alamb are we all ok with this design? I plan to update this PR tomorrow.

**Comment by alamb** [comment]
Hi @xinlifoobar  -- I haven't had a chance to review this PR unfortunately. Perhaps @tinfoil-knight  has some time to review  the design as the original filer of https://github.com/apache/datafusion/issues/10853

File: datafusion/sql/src/statement.rs
Comment (tinfoil-knight) [inline datafusion/sql/src/statement.rs:1157]:
```suggestion
                Err(_) => return plan_err!("Unsupported Value {}", value[0]),
```

File: datafusion/sql/src/utils.rs
Comment (tinfoil-knight) [reply datafusion/sql/src/utils.rs:290]: This function is also used in the match case for `SQLExpr::Value(v)` and even though we're returning `"Unsupported Value {}"` error there to users, this error message might cause confusion for some contributor stepping through the code.

**Review by tinfoil-knight — CHANGES_REQUESTED** [review]
PR looks good overall. Some minor changes requested. I prefer the current implementation that changes the normalization for all option values.

@berkaysynnada 's solution allows option-specific normalization but I don't think that kind of fine-grained control over option values is needed.

---
Sidenote:
We could add traits later on that'd allow users specify custom normalization behavior for identifiers and option values through `SqlToRel::new_with_options`.

```rust
// Example
pub trait Normalizer {
	fn normalize(&self, key: &str, value: Value) -> Result<String>;
}
```

**Review by tinfoil-knight — APPROVED** [review]

**Comment by tinfoil-knight** [comment]
@xinlifoobar Could you cleanup the commit history? Avoid using rebase. Just create a merge commit (if really necessary) w/ the `main` branch.

File: datafusion/sql/src/statement.rs
Comment (xinlifoobar) [reply datafusion/sql/src/statement.rs:1024]: Sorrry... done

File: datafusion/sql/src/planner.rs
Comment (tinfoil-knight) [inline datafusion/sql/src/planner.rs:118]:
```suggestion
    pub fn normalize(&self, value: String) -> String {
        if self.normalize {
            value.to_ascii_lowercase()
        } else {
            value
        }
    }
}
```

I don't think the normalize function should return an error. How about we do the string conversion and normalization in separate steps in the `parse_options_map` method. 

Something like:
```rs
let value_string = match value_to_string(&value) {
  Some(s) => self.value_normalizer.normalize(s),
  None => return plan_err!("Unsupported Value in Options : {}", value),
};
```

File: datafusion/sql/src/utils.rs
Comment (tinfoil-knight) [inline datafusion/sql/src/utils.rs:266]: `utils` is generally for things that are used across multiple files repeatedly.

How about we remove this function and just do the normalization in the `normalize` method under the `ValueNormalizer` struct?

If we separate the string conversion and normalization (as suggested in another comment), then we can move back the `value_to_string` fn too.

File: datafusion/sql/src/statement.rs
Comment (tinfoil-knight) [inline datafusion/sql/src/statement.rs:861]:
```suggestion
        let options_map = self.parse_options_map(options, true)?;
```

File: datafusion/sql/src/utils.rs
Comment (tinfoil-knight) [reply datafusion/sql/src/utils.rs:266]: Also, `exec_err` is only for Execution Errors. Code running in the planning stage should ideally use `plan_err`.

**Review by tinfoil-knight — CHANGES_REQUESTED** [review]

File: datafusion/sql/src/statement.rs
Comment (xinlifoobar) [reply datafusion/sql/src/statement.rs]: Done.

File: datafusion/sql/src/utils.rs
Comment (xinlifoobar) [reply datafusion/sql/src/utils.rs]: This follows the origin code structure used by `normalize_ident`. However, I have removed this though.

File: datafusion/sql/src/planner.rs
Comment (xinlifoobar) [reply datafusion/sql/src/planner.rs]: Done.

**Review by tinfoil-knight — APPROVED** [review]
@alamb The current design and PR looks good to me. Can we get this merged?

**Review by alamb — APPROVED** [review]
I agree -- this looks great to me. Than you @xinlifoobar  and @tinfoil-knight 

I'll plan to merge this tomorrow unless there are other comments.

cc @berkaysynnada

**Review by ozankabak — COMMENTED** [review]
> @berkaysynnada 's solution allows option-specific normalization but I don't think that kind of fine-grained control over option values is needed.

We actually do need this in our use case. Let's do the general solution to avoid API churn on such a user-visible part of DataFusion.

@xinlifoobar and @tinfoil-knight, do you have time to apply @berkaysynnada's suggestion? It's OK if not, we can work on it instead.

cc @berkaysynnada

**Comment by xinlifoobar** [comment]
> > @berkaysynnada 's solution allows option-specific normalization but I don't think that kind of fine-grained control over option values is needed.
> 
> We actually do need this in our use case. Let's do the general solution to avoid API churn on such a user-visible part of DataFusion.
> 
> @xinlifoobar and @tinfoil-knight, do you have time to apply @berkaysynnada's suggestion? It's OK if not, we can work on it instead.
> 
> cc @berkaysynnada

Hey, since this is pending long. I could do this as a follow up after this is merged.

**Comment by ozankabak** [comment]
Sure -- let's just make sure to avoid API churn and complete the work 🙂

**Comment by alamb** [comment]
> Sure -- let's just make sure to avoid API churn and complete the work 🙂

How about we file a ticket explaining what else is needed prior to merging this PR. I can do this but I want to make sure I have the use case down correctly

@berkaysynnada  says:

> What I suggest is that we can let people to define their own normalizations. Rather than directing all values to lowercase, people can declare their own normalization methods for specific option keys. If we choose to do that in this PR, I believe the normalization process should be done during the assignment of table options.

The code in https://github.com/synnada-ai/datafusion-upstream/commit/341b48458c09f25d2a5873fcee2fce18eb0e1097 I think seems to show a way to add an annotation to a built in config setting value that controls how it will be normalized. 

Is the use case to 
1. Provide a new set of `ConfigOptions` (not in DataFusion) that have custom normalization rules? 
2.  Avoid normalization on some of the built in configs?

Something else?

**Comment by ozankabak** [comment]
Both. We would like to have key-level control on what normalization means for config options, built-in and extension.

**Comment by alamb** [comment]
> Both. We would like to have key-level control on what normalization means for config options, built-in and extension.

In the interest of keeping the code moving, I filed  https://github.com/apache/datafusion/issues/11650 to track the additional feature and I merged up from main to resolve a conflict

**Comment by ozankabak** [comment]
Sounds good -- this is good to go from my perspective.

**Comment by alamb** [comment]
Clippy failure seems unrelated to this PR. I think it is https://github.com/apache/datafusion/issues/11651

**Comment by alamb** [comment]
Merged up to get the clippy fix

**Comment by alamb** [comment]
🚀  thanks and sorry for the delay @xinlifoobar

**Pull request merged**
